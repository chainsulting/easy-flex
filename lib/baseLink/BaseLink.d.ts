import React, { AnchorHTMLAttributes } from 'react';
import { Color } from '../types';
import { ColorProps } from '../utils/color';
import { FlexItemProps } from '../utils/flexItem';
import { FontProps } from '../utils/font';
import { MarginProps } from '../utils/margin';
import { SizeProps } from '../utils/size';
export interface BaseLinkProps extends Omit<AnchorHTMLAttributes<HTMLAnchorElement>, 'color'>, ColorProps, FlexItemProps, FontProps, MarginProps, SizeProps {
    /** Component's hover color. */
    hoverColor?: Color;
    /** If true, the link is opened in a new tab. */
    newTab?: boolean;
}
export declare const BaseLink: React.ForwardRefExoticComponent<BaseLinkProps & React.RefAttributes<HTMLAnchorElement>>;
