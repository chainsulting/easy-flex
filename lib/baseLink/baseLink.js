"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseLink = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var react_1 = require("react");
var styled_components_1 = __importDefault(require("styled-components"));
var utils_1 = require("../utils");
var StyledBaseLink = styled_components_1.default.a(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n\tbox-sizing: border-box;\n\ttext-decoration: none;\n\tpadding: 0;\n\talign-self: ", ";\n\tcolor: ", ";\n\tmargin-bottom: ", ";\n\tmargin-left: ", ";\n\tmargin-right: ", ";\n\tmargin-top: ", ";\n\n\t&:hover {\n\t\tcolor: ", ";\n\t}\n"], ["\n\tbox-sizing: border-box;\n\ttext-decoration: none;\n\tpadding: 0;\n\talign-self: ", ";\n\tcolor: ", ";\n\tmargin-bottom: ", ";\n\tmargin-left: ", ";\n\tmargin-right: ", ";\n\tmargin-top: ", ";\n\n\t&:hover {\n\t\tcolor: ", ";\n\t}\n"])), function (_a) {
    var alignSelf = _a["data-align-self"];
    return alignSelf;
}, function (_a) {
    var color = _a["data-color"];
    return color;
}, function (_a) {
    var marginBottom = _a["data-margin-bottom"];
    return marginBottom;
}, function (_a) {
    var marginLeft = _a["data-margin-left"];
    return marginLeft;
}, function (_a) {
    var marginRight = _a["data-margin-right"];
    return marginRight;
}, function (_a) {
    var marginTop = _a["data-margin-top"];
    return marginTop;
}, function (_a) {
    var hoverColor = _a["data-hover-color"];
    return hoverColor;
});
var BaseLink = function (_a) {
    var alignSelf = _a.alignSelf, children = _a.children, _b = _a.color, color = _b === void 0 ? 'inherit' : _b, hoverColor = _a.hoverColor, margin = _a.margin, marginBottom = _a.marginBottom, marginLeft = _a.marginLeft, marginRight = _a.marginRight, marginTop = _a.marginTop, marginX = _a.marginX, marginY = _a.marginY, _c = _a.newTab, newTab = _c === void 0 ? false : _c, props = __rest(_a, ["alignSelf", "children", "color", "hoverColor", "margin", "marginBottom", "marginLeft", "marginRight", "marginTop", "marginX", "marginY", "newTab"]);
    var processedColor = (0, utils_1.useColor)(color, undefined);
    var processedHoverColor = (0, utils_1.useColor)(hoverColor, undefined);
    var target = (0, react_1.useMemo)(function () { return (newTab ? '_blank' : undefined); }, [newTab]);
    var distance = (0, utils_1.useDistance)({
        margin: margin,
        marginBottom: marginBottom,
        marginLeft: marginLeft,
        marginRight: marginRight,
        marginTop: marginTop,
        marginX: marginX,
        marginY: marginY,
    });
    return ((0, jsx_runtime_1.jsx)(StyledBaseLink, __assign({ "data-align-self": alignSelf, "data-color": processedColor, "data-hover-color": processedHoverColor, "data-margin-bottom": distance.margin.bottom, "data-margin-left": distance.margin.left, "data-margin-right": distance.margin.right, "data-margin-top": distance.margin.top, rel: "noopener noreferrer", target: target }, props, { children: children })));
};
exports.BaseLink = BaseLink;
var templateObject_1;
