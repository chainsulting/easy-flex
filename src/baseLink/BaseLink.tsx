import React, { AnchorHTMLAttributes, FC, useMemo } from 'react';
import styled from 'styled-components';
import { IAbsoluteSize, IAlignSelf, IColor, ICssColor, IDistance } from '../types';
import { useColor, useDistance } from '../utils';

const StyledBaseLink = styled.a<{
	'data-align-self'?: IAlignSelf;
	'data-color'?: ICssColor;
	'data-hover-color'?: ICssColor;
	'data-margin-bottom': IAbsoluteSize;
	'data-margin-left': IAbsoluteSize;
	'data-margin-right': IAbsoluteSize;
	'data-margin-top': IAbsoluteSize;
}>`
	box-sizing: border-box;
	text-decoration: none;
	padding: 0;
	align-self: ${({ 'data-align-self': alignSelf }) => alignSelf};
	color: ${({ 'data-color': color }) => color};
	margin-bottom: ${({ 'data-margin-bottom': marginBottom }) => marginBottom};
	margin-left: ${({ 'data-margin-left': marginLeft }) => marginLeft};
	margin-right: ${({ 'data-margin-right': marginRight }) => marginRight};
	margin-top: ${({ 'data-margin-top': marginTop }) => marginTop};

	&:hover {
		color: ${({ 'data-hover-color': hoverColor }) => hoverColor};
	}
`;

export interface IBaseLinkProps extends AnchorHTMLAttributes<HTMLAnchorElement> {
	/** The alignment of the component on the parent's element cross axis. */
	alignSelf?: IAlignSelf;
	/** Component's color. */
	color?: IColor;
	/** Component's hover color. */
	hoverColor?: IColor;
	/** Component's margin of all sides. */
	margin?: IDistance | IAbsoluteSize;
	/** Component's bottom margin. */
	marginBottom?: IDistance | IAbsoluteSize;
	/** Component's left margin. */
	marginLeft?: IDistance | IAbsoluteSize;
	/** Component's right margin. */
	marginRight?: IDistance | IAbsoluteSize;
	/** Component's top margin. */
	marginTop?: IDistance | IAbsoluteSize;
	/** Component's left and right margin. */
	marginHorizontal?: IDistance | IAbsoluteSize;
	/** Component's top and bottom margin. */
	marginVertical?: IDistance | IAbsoluteSize;
	/** If true, the link is opened in a new tab. */
	newTab?: boolean;
}

export const BaseLink: FC<IBaseLinkProps> = ({
	alignSelf,
	children,
	color = 'inherit',
	hoverColor,
	margin,
	marginBottom,
	marginLeft,
	marginRight,
	marginTop,
	marginHorizontal,
	marginVertical,
	newTab = false,
	...props
}) => {
	const processedColor = useColor(color, undefined);

	const processedHoverColor = useColor(hoverColor, undefined);

	const target = useMemo<'_blank' | undefined>(() => (newTab ? '_blank' : undefined), [newTab]);

	const distance = useDistance({
		margin,
		marginBottom,
		marginLeft,
		marginRight,
		marginTop,
		marginHorizontal,
		marginVertical,
	});

	return (
		<StyledBaseLink
			data-align-self={alignSelf}
			data-color={processedColor}
			data-hover-color={processedHoverColor}
			data-margin-bottom={distance.margin.bottom}
			data-margin-left={distance.margin.left}
			data-margin-right={distance.margin.right}
			data-margin-top={distance.margin.top}
			rel="noopener noreferrer"
			target={target}
			{...props}
		>
			{children}
		</StyledBaseLink>
	);
};
