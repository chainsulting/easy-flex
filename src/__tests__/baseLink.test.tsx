import { render } from '@testing-library/react';
import 'jest-styled-components';
import React from 'react';
import { BaseLink } from '../baseLink/BaseLink';

describe('<BaseLink />', () => {
	test('should have the basic css settings', async () => {
		const result = render(<BaseLink />);
		expect(result.baseElement).toMatchSnapshot();
	});
});
