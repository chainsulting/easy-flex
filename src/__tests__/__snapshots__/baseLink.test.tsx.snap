// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BaseLink /> should have the basic css settings 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding: 0;
  color: inherit;
  margin-bottom: 0px;
  margin-left: 0px;
  margin-right: 0px;
  margin-top: 0px;
}

<body>
  <div>
    <a
      class="c0"
      data-color="inherit"
      data-margin-bottom="0px"
      data-margin-left="0px"
      data-margin-right="0px"
      data-margin-top="0px"
    />
  </div>
</body>
`;
